From 12da4b8f0c1e0ae1920810a1e333224e35dd5f42 Mon Sep 17 00:00:00 2001
From: Ido Yariv <ido@wizery.com>
Date: Mon, 2 Jan 2012 21:42:04 +0200
Subject: [PATCH 15/19] sdio: Add lock-less handlers host claiming function
 *EXPERIMENTAL*

Lock-less handlers which are called without the host being claimed
probably need to claim the host by themselves. In order to avoid a
race with sdio_release_irq, add support for claiming the host and
aborting the claim when the sdio irq is freed.

Signed-off-by: Eliad Peller <eliad@wizery.com>
---
 drivers/mmc/core/sdio_io.c    |   19 +++++++++++++++++++
 include/linux/mmc/sdio_func.h |    1 +
 2 files changed, 20 insertions(+), 0 deletions(-)

diff --git a/drivers/mmc/core/sdio_io.c b/drivers/mmc/core/sdio_io.c
index 549a341..86124e5 100755
--- a/drivers/mmc/core/sdio_io.c
+++ b/drivers/mmc/core/sdio_io.c
@@ -33,6 +33,25 @@ void sdio_claim_host(struct sdio_func *func)
 EXPORT_SYMBOL_GPL(sdio_claim_host);
 
 /**
+ *	sdio_claim_host_irq - exclusively claim a bus for a certain SDIO function
+ *	@func: SDIO function that will be accessed
+ *
+ *	Claim a bus for a set of operations. The SDIO function given
+ *	is used to figure out which bus is relevant.
+ *	The function may abort (returing a negative return value) if the sdio
+ *	irq is freed.
+ */
+int sdio_claim_host_irq(struct sdio_func *func)
+{
+	BUG_ON(!func);
+	BUG_ON(!func->card);
+
+	return __mmc_claim_host(func->card->host,
+				&func->card->host->sdio_irq_thread_abort);
+}
+EXPORT_SYMBOL_GPL(sdio_claim_host_irq);
+
+/**
  *	sdio_release_host - release a bus for a certain SDIO function
  *	@func: SDIO function that was accessed
  *
diff --git a/include/linux/mmc/sdio_func.h b/include/linux/mmc/sdio_func.h
index 5dff5e9..608c83b 100755
--- a/include/linux/mmc/sdio_func.h
+++ b/include/linux/mmc/sdio_func.h
@@ -126,6 +126,7 @@ extern void sdio_unregister_driver(struct sdio_driver *);
  * SDIO I/O operations
  */
 extern void sdio_claim_host(struct sdio_func *func);
+extern int sdio_claim_host_irq(struct sdio_func *func);
 extern void sdio_release_host(struct sdio_func *func);
 
 extern int sdio_enable_func(struct sdio_func *func);
-- 
1.7.7.6

